aead/src/lib.rs

```diff
@@ -104,22 +176,47 @@ pub trait AeadMut {
     ///
     /// See notes on [`Aead::encrypt()`] about allowable message payloads and
     /// Associated Additional Data (AAD).
+    #[cfg(feature = "alloc")]
     fn encrypt<'msg, 'aad>(
         &mut self,
         nonce: &GenericArray<u8, Self::NonceSize>,
         plaintext: impl Into<Payload<'msg, 'aad>>,
-    ) -> Result<Vec<u8>, Error>;
+    ) -> Result<Vec<u8>, Error> {
+        encrypt_to_postfix_tagged_vec!(self, nonce, plaintext)
+    }
+
+    /// Encrypt the data in-place, returning the authentication tag
+    fn encrypt_in_place_detached(
+        &mut self,
+        nonce: &GenericArray<u8, Self::NonceSize>,
+        associated_data: &[u8],
+        buffer: &mut [u8],
+    ) -> Result<GenericArray<u8, Self::TagSize>, Error>;
```

Comment:
Could you help me understand why you'd want to return the tag here, rather than doing that in-place as well? Other than requiring `split_at_mut()` in the macro it seems like it would otherwise be superior...

Reply:
That's actually what I originally suggested here:

https://github.com/RustCrypto/traits/pull/40#issuecomment-516878391

@newpavlov suggested returning the tag instead here: https://github.com/RustCrypto/traits/pull/40#issuecomment-516884094

I ended up liking that suggestion. I think it's cleaner because it's one less side effect, one less mutable argument, and actually leverages the return value.

Every time you rely on a side effect there's a risk it won't happen. That risk is unavoidable with an in-place API: mutation-by-side-effect is the name of the game. But we don't need to rely on it happening it twice, and can instead have the underlying implementation construct the tag by value, instead of it being one more thing for the caller to initialize with the hope the backing implementation actually does its job.

Reply:
I dunno, if the cipher implementation is doing all this in-place, and it has trouble mutating slices, my intuition is that you've probably got bigger problems than worrying about the tag getting written.

Reply:
I agree with @jcape on doing the tag in-place. Typically, the tag will be prefixed/suffixed to the ciphertext anyways. From a performance standpoint, it would be great if this didn't require a `memcpy`.

IMO, the risk of the backing implementation not mutating the tag is roughly the same as it returning a bad tag.

Reply:
This might be a bit extreme, but I would be in favor of having 0 visibility into the tag for **all** APIs. It should be treated as implicit ciphertext expansion.

Reply:
> This might be a bit extreme, but I would be in favor of having 0 visibility into the tag for **all** APIs. It should be treated as implicit ciphertext expansion.

There's an extensive discussion on the ciphertext layout in #40 :-) I'm generally in agreement with you (RFC and all), but if it prevents divergent implementations for weird non-RFC-compliant schemes like the encrypted filesystem thing where the tag lives in some separate metadata structure, I'm OK with exposing this API.